engine/ftactic.mli

```diff
@@ -43,7 +43,7 @@ val run : 'a t -> ('a -> unit Proofview.tactic) -> unit Proofview.tactic
 
 (** Enter a goal. The resulting tactic is focused. *)
 
-val enter : (Proofview.Goal.t -> 'a t) -> 'a t
+val enter : (Evd.evar_map -> Proofview.Goal.t -> 'a t) -> 'a t
```

Comment:
Maybe this would be a good moment to switch to a record:
```
module Info = struct
  type t =
    { sigma : Evd.evar_map
    ; gl : Goal.t
    }
end
```
Something I like about records in this position is that they tend to enforce a more uniform naming, that is to say `enter (fun { gl } -> ...)` is going always been to be more attractive than `enter (fun { gl = my_foo } -> ...)`

Reply:
The whole point is to not do that as the sigma keeps getting independently modified, so we would have to keep updating the record.

Reply:
I don't understand what you mean, sorry @SkySkimmer , the record is just a view on the enter data, so indeed why would you have to update it, other than calling refine?

Reply:
I guess if no API uses the record it's ok, the kind of problem we get currently is like 
~~~ocaml
enter (fun g -> 
  let sigma, t = pf_type_of g bla in 
  let sigma, t' = pf_type_of g bla' in (* !!! constraints for bla lost *)
~~~

Reply:
Yup, IMHO the record should only be a view for now [the question of a lower-level refine still stands tho]

I guess the main point on using a record or not is forward compatibility. Do you think enter may gain more parameters? If so, that could help, otherwise is more of a taste thing I guess.

Reply:
> I guess if no API uses the record it's ok, the kind of problem we get currently is like
> 
> ```ocaml
> enter (fun g -> 
>   let sigma, t = pf_type_of g bla in 
>   let sigma, t' = pf_type_of g bla' in (* !!! constraints for bla lost *)
> ```

Well, shouldn't `pf_type_of` be a tactic?

Reply:
FTR: https://github.com/coq/coq/blob/master/plugins/ssr/ssrcommon.mli#L428

Bet beware that two kind of tactics exist, the mono and multi goal one, but they have the same type! (so basically you have no clue what are doing when you are in the monad).

Reply:
After this PR both the `tac` and `tcl` version should commute, right?

Reply:
As far as I understand this PR exposes to the `enter` API the internal state of the monad.
I'm not an expert of monads, but I believe they should hide that state to the user, and not expose it. Of course if you don't have all your operators "lifted into the monad" then you are screwed, so the right fix is, IMO, to put all APIs that need a sigma in the monad rather than proving another way to break the abstraction.

Wrt mono/multi goal tactics, I don't think this PR adds anything.

Reply:
After seeing actual use I think it does make sense to lift the reader state into the parameters of `enter`, on the other hand, I do agree that writes should be always be a tactic.

Reply:
After reading the comments below I take back my comment above, indeed it is too easy for the lifted `get` to be shadowed.